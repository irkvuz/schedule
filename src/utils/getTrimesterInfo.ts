import { differenceInWeeks, format, parseISO } from 'date-fns';
import { ITrimesterOld } from '../constants';
import { getWeekNumber } from './getWeekNumber';

export interface ITrimesterInfo {
  weekNumberInSemester: number;
  weekNumberInYear: number;
  weeksInSemester: number;
  weeksInYear: number;
  dateStartString: string;
  dateFinishString: string;
}

export function getTrimesterInfo(
  trimester: ITrimesterOld,
  today: Date
): ITrimesterInfo {
  const dateStart = parseISO(trimester.dateStart);
  const dateFinish = parseISO(trimester.dateFinish);

  const weekNumberInSemester = getWeekNumber(
    trimester.dateStart,
    today
  );
  const weeksInSemester = differenceInWeeks(dateFinish, dateStart) + 1;
  /**
   * In second semestr it shoud be equal to number of weeks in prev simester
   */
  const weeksInPrevSemester = 15;
  const weekNumberInYear = weekNumberInSemester + weeksInPrevSemester;
  const weeksInYear = weeksInSemester + weeksInPrevSemester;

  const dateStartString= format(dateStart, 'dd.MM.yyyy');
  const dateFinishString = format(dateFinish, 'dd.MM.yyyy');

  return {
    weekNumberInSemester,
    weekNumberInYear,
    weeksInSemester,
    weeksInYear,
    dateStartString,
    dateFinishString,
  };
}
