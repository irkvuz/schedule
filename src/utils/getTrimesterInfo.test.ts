import { getTrimesterInfo } from "./getTrimesterInfo";


describe('getTrimesterInfo', ()=>{
  const trimesters = require('../../public/data/trimesters.json');
  
  it('should return correct trimester info for the first semester', ()=>{
    expect(getTrimesterInfo(trimesters[0], new Date('2018-09-08'))).toMatchSnapshot();
  });
  
  it('should return correct trimester info for the second semester', ()=>{
    expect(getTrimesterInfo(trimesters[1], new Date('2019-01-18'))).toMatchSnapshot();
  });
})