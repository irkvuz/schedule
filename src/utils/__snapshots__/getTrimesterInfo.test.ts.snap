// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correct trimester info 1`] = `
Object {
  "dateFinishString": "15.12.2018",
  "dateStartString": "01.09.2018",
  "weekNumberInSemester": 2,
  "weekNumberInYear": 17,
  "weeksInSemester": 15,
  "weeksInYear": 30,
}
`;
